File: src/sub/bendray.c
#include <math.h>
#include <stdio.h>

extern int icoordsystem;
extern const char *topo1file;
extern const char *topo2file;
extern void chtop(double x, double y, double *z, const char *topo1, const char *topo2);

void bendray(double rp[3][200], int nrp, const char *staname, double vtop, double *ttt) {
    if (icoordsystem != 2) {
        return;
    }

    double xydist = sqrtf((rp[0][nrp - 1] - rp[0][0]) * (rp[0][nrp - 1] - rp[0][0]) +
                         (rp[1][nrp - 1] - rp[1][0]) * (rp[1][nrp - 1] - rp[1][0]));
    if (xydist > 10.0f) {
        return;
    }

    double takeoff_num = sqrtf((rp[0][1] - rp[0][0]) * (rp[0][1] - rp[0][0]) +
                              (rp[1][1] - rp[1][0]) * (rp[1][1] - rp[1][0]));
    double takeoff_den = sqrtf((rp[0][1] - rp[0][0]) * (rp[0][1] - rp[0][0]) +
                              (rp[1][1] - rp[1][0]) * (rp[1][1] - rp[1][0]) +
                              (rp[2][1] - rp[2][0]) * (rp[2][1] - rp[2][0]));
    double takeoff_angle = 57.296f * asinf(takeoff_num / takeoff_den);
    if ((rp[2][1] - rp[2][0]) < 0.0f) {
        takeoff_angle = 180.0f - takeoff_angle;
    }

    int nrpm1 = nrp - 2;
    double arrive_num = sqrtf((rp[0][nrp - 1] - rp[0][nrpm1]) * (rp[0][nrp - 1] - rp[0][nrpm1]) +
                             (rp[1][nrp - 1] - rp[1][nrpm1]) * (rp[1][nrp - 1] - rp[1][nrpm1]));
    double arrive_den = sqrtf((rp[0][nrp - 1] - rp[0][nrpm1]) * (rp[0][nrp - 1] - rp[0][nrpm1]) +
                             (rp[1][nrp - 1] - rp[1][nrpm1]) * (rp[1][nrp - 1] - rp[1][nrpm1]) +
                             (rp[2][nrp - 1] - rp[2][nrpm1]) * (rp[2][nrp - 1] - rp[2][nrpm1]));
    double arrive_angle = 57.296f * asinf(arrive_num / arrive_den);
    if ((rp[2][1] - rp[2][0]) < 0.0f) {
        arrive_angle = 180.0f - arrive_angle;
    }

    int hypocintop = 0;
    double deltat1 = 0.0f;
    double deltat2 = 0.0f;

    double dx1 = 0.0f, dy1 = 0.0f, dz1 = 0.0f, xyz1 = 0.0f;
    if (rp[2][0] < 0.0f) {
        hypocintop = 1;
        dx1 = rp[0][1] - rp[0][0];
        dy1 = rp[1][1] - rp[1][0];
        dz1 = rp[2][1] - rp[2][0];
        xyz1 = sqrtf(dx1 * dx1 + dy1 * dy1 + dz1 * dz1);
    }

    if (hypocintop && takeoff_angle <= 45.0f) {
        hypocintop = 0;
    }
    if (!hypocintop && arrive_angle >= 135.0f) {
        return;
    }

    double dx2 = rp[0][nrp - 1] - rp[0][nrpm1];
    double dy2 = rp[1][nrp - 1] - rp[1][nrpm1];
    double dz2 = rp[2][nrp - 1] - rp[2][nrpm1];
    double xyz2 = sqrtf(dx2 * dx2 + dy2 * dy2 + dz2 * dz2);

    if (hypocintop) {
        dx1 /= 4.0f;
        dy1 /= 4.0f;
        dz1 /= 4.0f;
    }
    dx2 /= 4.0f;
    dy2 /= 4.0f;
    dz2 /= 4.0f;

    double rpn[3][10] = {{0.0f}};

    if (hypocintop) {
        for (int j = 0; j < 4; ++j) {
            rpn[0][j] = rp[0][0] + j * dx1;
            rpn[1][j] = rp[1][0] + j * dy1;
            rpn[2][j] = rp[2][0] + j * dz1;
        }
        rpn[0][4] = rp[0][1];
        rpn[1][4] = rp[1][1];
        rpn[2][4] = rp[2][1];
    }

    for (int j = 5; j <= 8; ++j) {
        rpn[0][j] = rp[0][nrpm1] + (j - 5) * dx2;
        rpn[1][j] = rp[1][nrpm1] + (j - 5) * dy2;
        rpn[2][j] = rp[2][nrpm1] + (j - 5) * dz2;
    }
    rpn[0][9] = rp[0][nrp - 1];
    rpn[1][9] = rp[1][nrp - 1];
    rpn[2][9] = rp[2][nrp - 1];

    if (hypocintop) {
        for (int j = 1; j <= 4; ++j) {
            double zzz = 0.0f;
            chtop(-rpn[0][j], rpn[1][j], &zzz, topo1file, topo2file);
            zzz += 0.1f;
            if (rpn[2][j] < zzz) {
                rpn[2][j] = zzz;
            }
        }
    }

    for (int j = 5; j <= 8; ++j) {
        double zzz = 0.0f;
        chtop(-rpn[0][j], rpn[1][j], &zzz, topo1file, topo2file);
        zzz += 0.1f;
        if (rpn[2][j] < zzz) {
            rpn[2][j] = zzz;
        }
    }

    double ttt1new = 0.0f;
    if (hypocintop) {
        for (int j = 1; j <= 4; ++j) {
            int jm1 = j - 1;
            double xyz1n = sqrtf((rpn[0][j] - rpn[0][jm1]) * (rpn[0][j] - rpn[0][jm1]) +
                               (rpn[1][j] - rpn[1][jm1]) * (rpn[1][j] - rpn[1][jm1]) +
                               (rpn[2][j] - rpn[2][jm1]) * (rpn[2][j] - rpn[2][jm1]));
            ttt1new += xyz1n / vtop;
        }
    }

    double ttt2new = 0.0f;
    for (int j = 6; j <= 9; ++j) {
        int jm1 = j - 1;
        double xyz2n = sqrtf((rpn[0][j] - rpn[0][jm1]) * (rpn[0][j] - rpn[0][jm1]) +
                           (rpn[1][j] - rpn[1][jm1]) * (rpn[1][j] - rpn[1][jm1]) +
                           (rpn[2][j] - rpn[2][jm1]) * (rpn[2][j] - rpn[2][jm1]));
        ttt2new += xyz2n / vtop;
    }

    if (hypocintop) {
        double ttt1old = xyz1 / vtop;
        deltat1 = ttt1new - ttt1old;
    }

    double ttt2old = xyz2 / vtop;
    deltat2 = ttt2new - ttt2old;

    if (fabs(deltat1) > 1e-5f || fabs(deltat2) > 1e-5f) {
        fprintf(stdout, " BENDRAY>>> ray bended below surface! Station: %.4s\n", staname);
        fprintf(stdout, " dt1=%6.3f   dt2=%6.3f\n", deltat1, deltat2);
    }

    *ttt += deltat1 + deltat2;
}

File: src_fortran/sub/bendray.f
c
c
c
c
c
c
      subroutine BENDRAY(rp,nrp,staname,vtop,ttt)
c
      implicit none
      include '../inc/vel_com.inc'
c
      real rp(3,inrpmax)
      integer nrp
      real vtop,ttt
      character*4 staname
c
      logical hypocintop
      real rpn(3,10)
      real xydist,takeoff_angle,arrive_angle
      real deltat1,deltat2,dx1,dy1,dz1,xyz1
      real dx2,dy2,dz2,xyz2,zzz,ttt1new,xyz1n
      real ttt2new,xyz2n,ttt1old,ttt2old
      integer nrpm1,jnrp,jnrpm1
c
      if(icoordsystem.ne.2) RETURN   ! topo-array only for switzerland available
c
      xydist=
     &   SQRT( (rp(1,nrp)-rp(1,1))**2
     &        +(rp(2,nrp)-rp(2,1))**2 )
      if(xydist.gt.10.0) RETURN !epicentral-distance too big for 'airy' rays!(?)
c
      takeoff_angle=
     &   SQRT( (rp(1,2)-rp(1,1))**2
     &        +(rp(2,2)-rp(2,1))**2 )
     &   /
     &   SQRT( (rp(1,2)-rp(1,1))**2
     &        +(rp(2,2)-rp(2,1))**2
     &        +(rp(3,2)-rp(3,1))**2 )
      takeoff_angle=57.296*ASIN(takeoff_angle)
      if( (rp(3,2)-rp(3,1)) .lt. 0.0 )then
         takeoff_angle=180.-takeoff_angle     ! ray is going upwards
      endif
c
      arrive_angle=
     &   SQRT( (rp(1,nrp)-rp(1,(nrp-1)))**2
     &        +(rp(2,nrp)-rp(2,(nrp-1)))**2 )
     &   /
     &   SQRT( (rp(1,nrp)-rp(1,(nrp-1)))**2
     &        +(rp(2,nrp)-rp(2,(nrp-1)))**2
     &        +(rp(3,nrp)-rp(3,(nrp-1)))**2 )
      arrive_angle=57.296*ASIN(arrive_angle)
      if( (rp(3,2)-rp(3,1)) .lt. 0.0 )then
         arrive_angle=180.-arrive_angle     ! ray is coming upwards
      endif
c
c     write(6,'(1x,''takeoff_angle='',f6.1,''  arrive_angle='',
c    &          f6.1)') takeoff_angle, arrive_angle
c
      deltat1=0.0
      deltat2=0.0
      hypocintop=.false.
      if(rp(3,1).lt.0.0)then
         hypocintop=.true.
c        compute XYZ1: the length between the first 2 raypoints
         dx1=rp(1,2)-rp(1,1)
         dy1=rp(2,2)-rp(2,1)
         dz1=rp(3,2)-rp(3,1)
         xyz1=SQRT(dx1**2 + dy1**2 + dz1**2)
      endif
c
c     if takeoff_angle at surface < 45  --> ray too steep to be 'airy'
c
      if(hypocintop.and.takeoff_angle.le.45.0) hypocintop=.false.
      if(hypocintop)then
         continue   ! stay here; ray-start could be in the air...
      else
         if(arrive_angle.ge.135.0) RETURN ! ray too steep to be 'airy'
      endif
c
      nrpm1=nrp-1
c     compute XYZ2: the length between the last 2 raypoints
      dx2=rp(1,nrp)-rp(1,nrpm1)
      dy2=rp(2,nrp)-rp(2,nrpm1)
      dz2=rp(3,nrp)-rp(3,nrpm1)
      xyz2=SQRT(dx2**2 + dy2**2 + dz2**2)
c
c     average (3D-)ray-length in the top-layer is about 2 kilometers;
c     now split ray in top-layer into 4 parts of about equal length:
      if(hypocintop)then
         dx1=dx1/4.
         dy1=dy1/4.
         dz1=dz1/4.
      endif
      dx2=dx2/4.
      dy2=dy2/4.
      dz2=dz2/4.
      if(hypocintop)then
         do jnrp=1,4  ! first raypoint = first old raypoint!
            rpn(1,jnrp)=rp(1,1)+(jnrp-1)*dx1
            rpn(2,jnrp)=rp(2,1)+(jnrp-1)*dy1
            rpn(3,jnrp)=rp(3,1)+(jnrp-1)*dz1
         enddo
         jnrp=5   ! set last raypoint to second old raypoint to keep accuracy!
         rpn(1,jnrp)=rp(1,2)
         rpn(2,jnrp)=rp(2,2)
         rpn(3,jnrp)=rp(3,2)
      endif
      do jnrp=6,9  ! first raypoint = second-last old raypoint!
         rpn(1,jnrp)=rp(1,nrpm1)+(jnrp-6)*dx2
         rpn(2,jnrp)=rp(2,nrpm1)+(jnrp-6)*dy2
         rpn(3,jnrp)=rp(3,nrpm1)+(jnrp-6)*dz2
      enddo
      jnrp=10 ! set last raypoint to last old raypoint to keep accuracy!
      rpn(1,jnrp)=rp(1,nrp)
      rpn(2,jnrp)=rp(2,nrp)
      rpn(3,jnrp)=rp(3,nrp)
c
c     now set raypoints min. 100meters below surface (if necessary):
c
      if(hypocintop)then
         do jnrp=2,5
            call CHTOP(-rpn(1,jnrp),rpn(2,jnrp),zzz,
     &                 topo1file,topo2file)
            zzz=zzz+0.1  ! 100meters below surface
            if(rpn(3,jnrp).lt.zzz) rpn(3,jnrp)=zzz  ! set hypoc. below surface !
         enddo
      endif
      do jnrp=6,9
         call CHTOP(-rpn(1,jnrp),rpn(2,jnrp),zzz,
     &              topo1file,topo2file)
         zzz=zzz+0.1  ! 100meters below surface
         if(rpn(3,jnrp).lt.zzz) rpn(3,jnrp)=zzz   ! set hypoc. below surface !
      enddo
c
c     calculate ray-length-difference and correct traveltime:
c
      ttt1new=0.0
      if(hypocintop)then
         do jnrp=2,5
            jnrpm1=jnrp-1
            xyz1n=SQRT( (rpn(1,jnrp)-rpn(1,jnrpm1))**2
     &                 +(rpn(2,jnrp)-rpn(2,jnrpm1))**2
     &                 +(rpn(3,jnrp)-rpn(3,jnrpm1))**2 )
            ttt1new=ttt1new+xyz1n/vtop
         enddo
      endif
      ttt2new=0.0
      do jnrp=7,10
         jnrpm1=jnrp-1
         xyz2n=SQRT( (rpn(1,jnrp)-rpn(1,jnrpm1))**2
     &              +(rpn(2,jnrp)-rpn(2,jnrpm1))**2
     &              +(rpn(3,jnrp)-rpn(3,jnrpm1))**2 )
         ttt2new=ttt2new+xyz2n/vtop
      enddo
      if(hypocintop)then
         ttt1old=xyz1/vtop
         deltat1=(ttt1new-ttt1old)
      endif
      ttt2old=xyz2/vtop
      deltat2=(ttt2new-ttt2old)
      if(ABS(deltat1).gt.1e-5.or.ABS(deltat2).gt.1e-5)then
         write(6,'(1x,''BENDRAY>>> ray bended below surface!'',
     &                '' Station: '',a4)') staname
         write(6,'(1x,''dt1='',f6.3,''   dt2='',f6.3)') deltat1,deltat2
      endif
      ttt=ttt+deltat1+deltat2
c
      RETURN
      end ! of subr. bendray
